python/pyspark/mllib/linalg/distributed.py

```diff
@@ -0,0 +1,513 @@
+        elif class_name == "IndexedRowMatrix":
+            # We use DataFrames for serialization of IndexedRows from
+            # Java, so we first convert the RDD of rows to a DataFrame
+            # on the Scala/Java side. Then we map each Row in the
+            # DataFrame back to an IndexedRow on this side.
+            rows_df = callMLlibFunc("getIndexedRows", java_matrix)
+            rows = rows_df.map(lambda row: IndexedRow(row[0], row[1]))
+            return IndexedRowMatrix(rows, java_matrix=java_matrix)
+        elif class_name == "CoordinateMatrix":
+            # We use DataFrames for serialization of MatrixEntry entries
+            # from Java, so we first convert the RDD of entries to a
+            # DataFrame on the Scala/Java side. Then we map each Row in
+            # the DataFrame back to a MatrixEntry on this side.
+            entries_df = callMLlibFunc("getMatrixEntries", java_matrix)
+            entries = entries_df.map(lambda row: MatrixEntry(row[0], row[1], row[2]))
+            return CoordinateMatrix(entries, java_matrix=java_matrix)
+        else:
+            raise TypeError("Cannot create distributed matrix from Java %s" % class_name)
+    else:
+        raise TypeError("java_matrix should be JavaObject, got %s" % type(java_matrix))
+
+
+class DistributedMatrix(object):
+    """
+    Represents a distributively stored matrix backed by one or
+    more RDDs.
+
+    """
+    def numRows(self):
+        """Get or compute the number of rows."""
+        raise NotImplementedError
+
+    def numCols(self):
+        """Get or compute the number of cols."""
+        raise NotImplementedError
+
+
+class RowMatrix(DistributedMatrix):
+    """
+    .. note:: Experimental
+
+    Represents a row-oriented distributed Matrix with no meaningful
+    row indices.
+
+    :param rows: An RDD of vectors.
+    :param numRows: Number of rows in the matrix. A non-positive
+                    value means unknown, at which point the number
+                    of rows will be determined by the number of
+                    records in the `rows` RDD.
+    :param numCols: Number of columns in the matrix. A non-positive
+                    value means unknown, at which point the number
+                    of columns will be determined by the size of
+                    the first row.
+    """
+    def __init__(self, rows, numRows=0, numCols=0, java_matrix=None):
+        """Create a wrapper over a Java RowMatrix."""
+        if not isinstance(rows, RDD):
+            raise TypeError("rows should be an RDD of vectors, got %s" % type(rows))
+        rows = rows.map(_convert_to_vector)
+
+        if not (isinstance(java_matrix, JavaObject)
+                and java_matrix.getClass().getSimpleName() == "RowMatrix"):
+            java_matrix = callMLlibFunc("createRowMatrix", rows, long(numRows), int(numCols))
+
+        self._java_matrix_wrapper = JavaModelWrapper(java_matrix)
+        self._rows = rows
```

Comment:
Yeah we're definitely keeping a `@property def rows` still, and at a minimum, computing the rows from the Java side when it is called. It would be nice to not waste computation if we already had an RDD to start with on the Python side, but I think the implementation could get messy if we keep checking whether or not rows is an RDD or a JavaObject everywhere.